### PR TITLE
add imageGcWhitelist  codes that are metioned in issue #66221

### LINKED ISSUE #66221: shall we add a image gc white list that the kubelet imge gc will ignore?
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What happened**:
In some offline environment,we will load some images for furture use , so if the  disk usage exceeds the image-gc-high-threshold ,kubelet mya delete it, because of the offline, so the pod will never start

we use kubernetes as a basci platform, so we do not have a image registry

**What you expected to happen**:
Shall we add a function, that put the image in a whitelist ,so the image gc will ignore it ?
We had finished the code and test it ,if we need it ,we can do a pull request.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by zhangmingld**
**What this PR does / why we need it**:
As the issue mentioned，in some offline case ， when we pre-load some dokcer images ，kubelet may delete it because of the exhausted resource，so we can not use the docker image when needed

In this commit ,people can add a whitelistimage using --image-gc-whitelist=imagename:tag... or a kubelet config file or dynamic config

we have tested it for a long time ,and released it in our production

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
fixes #66221 

**Special notes for your reviewer**:

**Release note**:

```release-note
add a imageGCWhitelist so that the image gc will ignore
```

### CODE REVIEW

**Comment by zhangmingld** [comment]
@sig-node-reviewers
can you please help review this?

**Comment by senthilrch** [comment]
##68930

**Comment by zhangmingld** [comment]
@sjenning @Random-Liu 
Can you help me review this?
Thank you~

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -506,6 +506,7 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
 	fs.StringSliceVar(&c.ClusterDNS, "cluster-dns", c.ClusterDNS, "Comma-separated list of DNS server IP address.  This value is used for containers DNS server in case of Pods with \"dnsPolicy=ClusterFirst\". Note: all DNS servers appearing in the list MUST serve the same set of records otherwise name resolution within the cluster may not work correctly. There is no guarantee as to which DNS server may be contacted for name resolution.")
 	fs.DurationVar(&c.StreamingConnectionIdleTimeout.Duration, "streaming-connection-idle-timeout", c.StreamingConnectionIdleTimeout.Duration, "Maximum time a streaming connection can be idle before the connection is automatically closed. 0 indicates no timeout. Example: '5m'")
 	fs.DurationVar(&c.NodeStatusUpdateFrequency.Duration, "node-status-update-frequency", c.NodeStatusUpdateFrequency.Duration, "Specifies how often kubelet posts node status to master. Note: be cautious when changing the constant, it must work with nodeMonitorGracePeriod in nodecontroller.")
+	fs.StringSliceVar(&c.ImageGCWhitelist, "image-gc-whitelist", c.ImageGCWhitelist, "Comma-separated list of image which will be ignored by image garbage collection")
```
Comment (Monkeyanator) [inline cmd/kubelet/app/options/options.go]:
```suggestion
	fs.StringSliceVar(&c.ImageGCWhitelist, "image-gc-whitelist", c.ImageGCWhitelist, "Comma-separated list of images which will be ignored by image garbage collection")
```

**Comment by zhangmingld** [comment]
@Monkeyanator Thankyou , i have commit the code

**Comment by libesz** [comment]
Is there any string size constraint which can limit the image list length?

**Comment by libesz** [comment]
> 
> 
> > Is there any string size constraint which can limit the image list length?
> 
> currently, we did not limit the num , will we need it in some scenarios?

I don't want to limit, just like to know whether any is built-in to the param processing. Assuming I have a relatively large number of preloaded images (whole container infra in VM image), the configured value for kubelet can be huge. I am thinking something like wide-cards or regexp possibilities in this parameter (infra/* or /infra\/.*/ ).

**Comment by libesz** [comment]
Could anyone push this forward?
/cc @sjenning @Random-Liu

**Comment by derekwaynecarr** [comment]
I think this feature is useful if its usage is well-controlled.  I think this feature could be queued up for 1.15.  It will need to go through the new API review process.  I am wondering if we can collect more detail on various use-cases people have for having a set of images ignored to know if this explicit whitelist was sufficient versus some other type of pattern match.

**Comment by szuecs** [comment]
@derekwaynecarr my use case would be to preload toolbox (CoreOS investigation container via systemd unit for example).
Another one can be to preload system daemonsets.

**Comment by mtaufen** [comment]
Agree with @derekwaynecarr. While this sounds like a good idea, it's a bit late in the release cycle to add this feature to 1.14. I think gathering feedback and targeting 1.15 would be a good target.

**Comment by zhangmingld** [comment]
Thankyou for all your reply , i will handle the conflict later

**Comment by zhangmingld** [comment]
Thankyou for all your reply ,i will handle the confilicts and api review later

And How about some whitlist key words ? In this way , we do not have to explicit the image , but some key words that can match a lot of images?

**Comment by libesz** [comment]
@zhanghuidinah My proposal (to keep things simple) would be to whitelist complete repos if the image name/tag is not specified. This could be a semi-dumb string matching (instead of having regexes in command line) to the beginning of the image reference.
Example: `--image-gc-whitelist infra/` matches to all images tagged to the fake remote repo, called `infra`.
This covers the image preload use-cases IMHO, where the embedded images cannot be pulled from anywhere after deploy. In this case, when one building the node image, the container images can be tagged to a common fake remote repo.

**Comment by mmpei** [comment]
maybe we could define a default repo like 'permanent-images/'. and all images in this repo will be ignored when doing image gc.
this will make things simple.

**Comment by libesz** [comment]
@zhangmingld Are you going to to create the KEP for this any soon? If no, I can offer our help to do it with the requirements and alternatives from this PR so far. Just to make sure we don't miss the next milestone.
/cc @csatarigergely

**Comment by szuecs** [comment]
@zhangmingld  @libesz  why not having a slice of regexp?
We do not need to be super efficient in this case (regexp vs string prefix match) and I would rather put a nil or empty slice as default. Maybe it is fine and it supports already changes in tags?

**Comment by libesz** [comment]
@szuecs Fine for me.

**Comment by zhangmingld** [comment]
> @zhangmingld Are you going to to create the KEP for this any soon? If no, I can offer our help to do it with the requirements and alternatives from this PR so far. Just to make sure we don't miss the next milestone.
> /cc @CsatariGergely

Thankyou very much. I think I can finish it in this week ~

**Comment by zhangmingld** [comment]
> regexp
Will regexp be a little complex for normal users ?
keywords or imagename:tag may be more intuitive to users

**Comment by libesz** [comment]
@zhangmingld @szuecs Yes regexp may be a little complicated, I agree. I would let the KEP reviewers to vote for the simpler or the regexp alternative. However, one can implement this in a way IMHO, where explicit imagename:tag can act as a regexp, which matches only a single image. 😉

**Comment by szuecs** [comment]
I don't think regexp is complicated for the audience.

**Comment by zhangmingld** [comment]
@libesz @derekwaynecarr Do we really need a KEP for this ?  I think it is  just a improvement for kubelet image management

**Comment by libesz** [comment]
@zhangmingld I personally don't need it 😄, because I could live with either solution. I asked for guidance on slack how to proceed... and that's what we got from @derekwaynecarr and @mtaufen. We have multiple options on what/how to implement it, and this also adds new CLI parameter to kubelet. I guess aligning on these requires the API KEP instead of just counting the reactions on the proposal comments. But others please correct me.

**Comment by CsatariGergely** [comment]
@zhangmingld we are ready to start working on a KEP. Is it okay if we create one and invite you for a review?

**Comment by zhangmingld** [comment]
> @zhangmingld we are ready to start working on a KEP. Is it okay if we create one and invite you for a review?

@CsatariGergely  Thankyou , that is ok , I m really busy these days

**Comment by CsatariGergely** [comment]
@zhangmingld [here](https://github.com/kubernetes/enhancements/pull/1007) is a link to the KEP pr.

**Comment by bluefriday** [comment]
Does the current version of kuberentes include this feature?

**Comment by zhangmingld** [comment]
i will reopen this as it is still needed

**Comment by mjnovice** [comment]
@zhangmingld could you help tell why this got closed ? We need something similar. :(

**Comment by szuecs** [comment]
@mjnovice the bot closed it

**Pull request closed without merging**